json/test_rapid_sax.cpp

```diff
@@ -21,83 +21,106 @@ struct coordinate_t {
   double y;
   double z;
 
-  auto operator<=>(const coordinate_t&) const = default;
+  auto operator<=>(const coordinate_t &) const = default;
 
-  friend ostream& operator<< (ostream &out, const coordinate_t &point) {
-    out << "coordinate_t {x: " << point.x
-        << ", y: " << point.y
+  friend ostream &operator<<(ostream &out, const coordinate_t &point) {
+    out << "coordinate_t {x: " << point.x << ", y: " << point.y
         << ", z: " << point.z << "}";
     return out;
   }
 };
 
-using TCallback = function<void(const coordinate_t&)>;
+using TCallback = function<void(const coordinate_t &)>;
 
 class CoordinateHandler : public BaseReaderHandler<UTF8<>, CoordinateHandler> {
 public:
-  CoordinateHandler(const TCallback& callback):
-    state_(kStart), x_(), y_(), z_(), callback_(callback) {}
+  CoordinateHandler(const TCallback &callback)
+      : state_(kStart), x_(), y_(), z_(), callback_(callback) {}
 
   bool Double(double d) {
     switch (state_) {
-      case kX: x_ += d; state_ = kCoordinatesElement; break;
-      case kY: y_ += d; state_ = kCoordinatesElement; break;
-      case kZ: z_ += d; state_ = kCoordinatesElement; break;
-      default: break;
+    case kX:
+      x_ += d;
+      state_ = kCoordinatesElement;
+      break;
+    case kY:
+      y_ += d;
+      state_ = kCoordinatesElement;
+      break;
+    case kZ:
+      z_ += d;
+      state_ = kCoordinatesElement;
+      break;
+    default:
+      break;
     }
     return true;
   }
 
   bool StartObject() {
     switch (state_) {
-      case kStart: state_ = kRoot; break;
-      case kCoordinatesArray: state_ = kCoordinatesElement; break;
-      default: break;
+    case kStart:
+      state_ = kRoot;
+      break;
+    case kCoordinatesArray:
+      state_ = kCoordinatesElement;
+      break;
+    default:
+      break;
     }
     return true;
   }
 
-  bool Key(const Ch* str, SizeType len, bool) {
+  bool Key(const Ch *str, SizeType len, bool) {
     switch (state_) {
-      case kRoot:
-        if (len == sizeof("coordinates") - 1 && memcmp(str, "coordinates", sizeof("coordinates") - 1) == 0)
-          state_ = kCoordinates;
-        break;
+    case kRoot:
+      if (len == sizeof("coordinates") - 1 &&
```

Comment:
This auto-formatting looks a little bit ugly too. Plus it seems a little bit hackish (I haven't noticed that before). I'd implement the following solution (based on https://stackoverflow.com/questions/2111667/compile-time-string-hashing/23684632 ).

Add hash functions (as it would allow to use us the `switch` statement), could go into `common/`:
```cpp
constexpr std::size_t cx_hash(const char *str, size_t len) {
  std::size_t hash = 0xcbf29ce484222325;
  const std::size_t prime = 0x00000100000001b3;

  for (size_t i = 0; i < len; ++i) {
    hash ^= static_cast<size_t>(str[i]);
    hash *= prime;
  }

  return hash;
}

std::size_t constexpr operator "" _hash(const char *str, size_t len) {
  return cx_hash(str, len);
}
```

and use it here:

```cpp
const auto hash = cx_hash(str, len);
if (hash == "coordinates"_hash) {
  state_ = kCoordinates;
}
```
Plus please see another comment below.

Reply:
One thing I like in benchmarks is compare the complexity of the codes, and too much code in `common/` will  hide it. At least, `notify()` isn't "part" of the implementations.

Reply:
`notify` is an exception because we need to compile  it as a part of dynamic library used for FFI with languages which don't support sockets out of box.

I guess we can copy paste blocks of code though (even for notify).